### PR TITLE
AABB API Implementation

### PR DESCRIPTION
**Pull request by DDoS**
https://github.com/SpongePowered/SpongeAPI/pull/1318

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/block/state/MixinStateImplementation.java [hunk 1]
```diff
@@ -122,7 +114,6 @@ public BlockSnapshot snapshotFor(Location<World> location) {
         }
         return builder.build();
     }
-
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/block/state/MixinStateImplementation.java]: why?

File: src/main/java/org/spongepowered/common/world/extent/SoftBufferExtentViewDownsize.java [hunk 2]
```diff
@@ -442,6 +443,30 @@ public boolean spawnEntity(Entity entity, Cause cause) {
     }
 
     @Override
+    public Set<Entity> getIntersectingEntities(AABB box, Predicate<Entity> filter) {
+        checkRange(box.getMin().getX(), box.getMin().getY(), box.getMin().getZ());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/world/extent/SoftBufferExtentViewDownsize.java]: Perhaps add a `checkRange` method which accepts a vec?

File: src/main/java/org/spongepowered/common/world/extent/ExtentViewTransform.java [hunk 3]
```diff
@@ -511,19 +511,65 @@ public Extent getExtentView(DiscreteTransform3 transform) {
...
+    }
+
+    @Override
+    public Optional<AABB> getBlockSelectionBox(int x, int y, int z) {
+        return this.extent.getBlockSelectionBox(this.inverseTransform.transformX(x, y, z), this.inverseTransform.transformY(x, y, z),
+            this.inverseTransform.transformZ(x, y, z));
+    }
+
+    @Override
+    public Set<Entity> getIntersectingEntities(AABB box, Predicate<Entity> filter) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/world/extent/ExtentViewTransform.java]: Should `checkNotNull` all of these methods

File: src/main/java/org/spongepowered/common/world/extent/ExtentViewDownsize.java [hunk 4]
```diff
@@ -496,23 +495,74 @@ public Extent getExtentView(DiscreteTransform3 transform) {
...
 
     @Override
+    public Optional<AABB> getBlockSelectionBox(int x, int y, int z) {
+        checkRange(x, y, z);
+        return this.extent.getBlockSelectionBox(x, y, z);
+    }
+
+    @Override
+    public Set<Entity> getIntersectingEntities(AABB box, Predicate<Entity> filter) {
+        checkRange(box.getMin().getX(), box.getMin().getY(), box.getMin().getZ());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/world/extent/ExtentViewDownsize.java]: See above comment regarding `checkRange`

File: src/main/java/org/spongepowered/common/mixin/core/block/state/MixinStateImplementation.java (hunk 1 above)
Comment (DDoS) [inline src/main/java/org/spongepowered/common/mixin/core/block/state/MixinStateImplementation.java]: I don't know, whatever.

File: src/main/java/org/spongepowered/common/world/extent/SoftBufferExtentViewDownsize.java (hunk 2 above)
Comment (DDoS) [inline src/main/java/org/spongepowered/common/world/extent/SoftBufferExtentViewDownsize.java]: Some other day.

File: src/main/java/org/spongepowered/common/util/VecHelper.java [hunk 5]
```diff
@@ -127,7 +129,25 @@ public static boolean inBounds(int x, int y, int z, Vector3i min, Vector3i max)
...
 
+    public static boolean inBounds(Vector3d pos, Vector3i min, Vector3i max) {
+        return inBounds(pos.getX(), pos.getY(), pos.getZ(), min, max);
+    }
+
     public static boolean inBounds(double x, double y, double z, Vector3i min, Vector3i max) {
         return x >= min.getX() && x <= max.getX() && y >= min.getY() && y <= max.getY() && z >= min.getZ() && z <= max.getZ();
     }
+
+    public static AxisAlignedBB toMC(AABB box) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/util/VecHelper.java]: `toNative` is what this method (sponge -> minecraft) is usually called.

Comment (DDoS) [inline src/main/java/org/spongepowered/common/util/VecHelper.java]: Doesn't really matter.

**Comment by DDoS** [comment]
Requesting permission to merge tonight.

**Pull request closed without merging**